src/InlineRelationsTransformer.cpp

```diff
@@ -0,0 +1,965 @@
+                    // Records of unequal size can't be equated
+                    return false;
+                }
+
+                // Equate all corresponding arguments
+                for (int i = 0; i < lhsArgs.size(); i++) {
+                    sub.push_back(std::make_pair(lhsArgs[i], rhsArgs[i]));
+                }
+
+                // Get rid of the record equality
+                sub.erase(sub.begin() + i);
+                done = false;
+            } else if ((isRecord(lhs) && isConstant(rhs)) || (isConstant(lhs) && isRecord(rhs))) {
+                // A record =/= a constant
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+// Returns the nullable vector of substitutions needed to unify the two given atoms.
+// If unification is not successful, the returned vector is marked as invalid.
+// Assumes that the atoms are both of the same relation.
+NullableVector<std::pair<AstArgument*, AstArgument*>> unifyAtoms(AstAtom* first, AstAtom* second) {
+    std::vector<std::pair<AstArgument*, AstArgument*>> substitution;
+
+    std::vector<AstArgument*> firstArgs = first->getArguments();
+    std::vector<AstArgument*> secondArgs = second->getArguments();
+
+    // Create the initial unification equalities
+    for (int i = 0; i < firstArgs.size(); i++) {
+        substitution.push_back(std::make_pair(firstArgs[i], secondArgs[i]));
+    }
+
+    // Reduce the substitutions
+    bool success = reduceSubstitution(substitution);
+    if (success) {
+        return NullableVector<std::pair<AstArgument*, AstArgument*>>(substitution);
+    } else {
+        // Failed to unify the two atoms
+        return NullableVector<std::pair<AstArgument*, AstArgument*>>();
+    }
+}
+
+// Inlines the given atom based on a given clause.
+// Returns the vector of replacement literals and the necessary constraints.
+// If unification is unsuccessful, the vector of literals is marked as invalid.
+std::pair<NullableVector<AstLiteral*>, std::vector<AstConstraint*>> inlineBodyLiterals(
+        AstAtom* atom, AstClause* atomInlineClause) {
+    bool changed = false;
+    std::vector<AstLiteral*> addedLits;
+    std::vector<AstConstraint*> constraints;
+
+    // Start off by renaming the variables in the inlined clause to avoid conflicts when unifying multiple
+    // atom
+    // - particularly when an inlined relation appears twice in a clause.
+    static int inlineCount = 0;
+
+    // Make a temporary clone so we can rename variables without fear
+    AstClause* atomClause = atomInlineClause->clone();
+
+    struct VariableRenamer : public AstNodeMapper {
+        int varnum;
+        VariableRenamer(int varnum) : varnum(varnum) {}
+        std::unique_ptr<AstNode> operator()(std::unique_ptr<AstNode> node) const override {
+            if (AstVariable* var = dynamic_cast<AstVariable*>(node.get())) {
+                // Rename the variable
+                AstVariable* newVar = var->clone();
```

Comment:
Just make the unique pointer to begin with, since you don't need the raw pointer.
`auto newVar = std::make_unique<AstVariable>(var->clone());`